pkg/registry/scheduling/workload/declarative_validation_test.go

```diff
@@ -0,0 +1,469 @@
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	apitesting "k8s.io/kubernetes/pkg/api/testing"
+	"k8s.io/kubernetes/pkg/apis/scheduling"
+
+	// Ensure all API groups are registered with the scheme
+	_ "k8s.io/kubernetes/pkg/apis/scheduling/install"
+)
+
+func TestDeclarativeValidate(t *testing.T) {
+	apiVersions := []string{"v1alpha1"} // Workload is currently only in v1alpha1
+	for _, apiVersion := range apiVersions {
+		t.Run(apiVersion, func(t *testing.T) {
+			testDeclarativeValidate(t, apiVersion)
+		})
+	}
+}
+
+func testDeclarativeValidate(t *testing.T, apiVersion string) {
+	ctx := genericapirequest.WithRequestInfo(genericapirequest.NewDefaultContext(), &genericapirequest.RequestInfo{
+		APIGroup:          "scheduling.k8s.io",
+		APIVersion:        apiVersion,
+		Resource:          "workloads",
+		IsResourceRequest: true,
+		Verb:              "create",
+	})
+
+	testCases := map[string]struct {
+		input        scheduling.Workload
+		expectedErrs field.ErrorList
+	}{
+		"valid": {
+			input: mkValidWorkload(),
+		},
+		"empty podGroups": {
+			input: mkValidWorkload(func(obj *scheduling.Workload) {
+				obj.Spec.PodGroups = nil
+			}),
+			expectedErrs: field.ErrorList{
+				field.Required(field.NewPath("spec", "podGroups"), "must have at least one item"),
+			},
+		},
+		"too many podGroups": {
+			input: mkValidWorkload(func(obj *scheduling.Workload) {
+				obj.Spec.PodGroups = make([]scheduling.PodGroup, scheduling.WorkloadMaxPodGroups+1)
+				for i := range obj.Spec.PodGroups {
+					obj.Spec.PodGroups[i] = scheduling.PodGroup{
+						Name: fmt.Sprintf("group-%d", i),
+						Policy: scheduling.PodGroupPolicy{
+							Gang: &scheduling.GangSchedulingPolicy{
+								MinCount: 1,
+							},
+						},
+					}
+				}
+			}),
+			expectedErrs: field.ErrorList{
+				field.TooMany(field.NewPath("spec", "podGroups"), scheduling.WorkloadMaxPodGroups+1, scheduling.WorkloadMaxPodGroups).WithOrigin("maxItems"),
+			},
+		},
+		"empty podGroup name": {
+			input: mkValidWorkload(func(obj *scheduling.Workload) {
+				obj.Spec.PodGroups[0].Name = ""
+			}),
+			expectedErrs: field.ErrorList{
+				field.Required(field.NewPath("spec", "podGroups").Index(0).Child("name"), ""),
+			},
+		},
+		"invalid podGroup name": {
+			input: mkValidWorkload(func(obj *scheduling.Workload) {
+				obj.Spec.PodGroups[0].Name = "Invalid_Name"
+			}),
+			expectedErrs: field.ErrorList{
+				field.Invalid(field.NewPath("spec", "podGroups").Index(0).Child("name"), "Invalid_Name", "a lowercase RFC 1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')").WithOrigin("format=k8s-short-name"),
+			},
+		},
+		"duplicate podGroup names": {
+			input: mkValidWorkload(func(obj *scheduling.Workload) {
+				obj.Spec.PodGroups = append(obj.Spec.PodGroups, scheduling.PodGroup{
```

Comment:
Consider encapsulating this into a function like:

```
func addPodGroup(name string) func(obj *scheduling.Workload) {
    return func(obj *scheduling.Workload) {
        obj.Spec.PodGroups = append(obj.Spec.PodGroups, scheduling.PodGroup{
            //... use `name`
        })
    }
}
```

So this test-case becomes:

```
		"duplicate podGroup names": {
            input: mkValidWorkload(addPodGroup("main")),
        expectedErrs: field.ErrorList{
            field.Duplicate(field.NewPath("spec.podGroups[1]"), nil))
        }
````

Note that I elided the value argument because it doesn't actually add clarity (and is not actually considered in the error matcher anyway!!).

Less LOCs, way more readable.